### DISCUSSION TITLE
Discussion: ent + metrics

### DISCUSSION THREAD

**Discussion by rotemtam** [ent/ent#1819] [General]
Opening a discussion following some comments on https://github.com/ent/contrib/pull/113

I will try to tl;dr; - we want to add to `ent/contrib` an extension to allow people to easily export metrics from their ent applications. The PR adds a `Hook` that can be used to instrument mutations via Prometheus.  Ideally, we would want to use OpenTelemetry but their repo states that the metrics part is currently on hold. OpenCensus states that it is being merged into OpenTelemetry. 

I don't think we should try to develop our own abstraction for exporting metrics but rather use a standards-based solution. I don't have a lot of experience with either OTEL or OC to make the decisions either way (adopt/reject/wait). Eventually, if no standards-based solution is relevant, we can create an interface that is the small subset of instrumentation features that we need for ent and develop metric backend implementations for Prometheus, Influx, CloudWatch, etc..  but again, that isn't great. 

So - opening up for discussion - how should we proceed? 

cc @a8m @yonidavidson @cliedeman @giautm

**Comment by cliedeman** [comment]
Hi @rotemtam I have used prometheus and opencencus (pre otel). I find there are enough small differences (especially with opentelemetry) that creating a shared interface would be difficult. So I think three libraries entprom, entopencensus, entotel would probably be the easiest. 

The hooks/operations code make it *really* easy to implement mutation metrics (The core code is ~100 lines). This allows anyone to easily implement metrics for their use case.
I am more interested in considering how to implement metrics for the read operations. Especially with edge loading.
